STIPS/STIP-004.md

```diff
@@ -0,0 +1,642 @@
+   </td>
+   <td>
+s * 1000
+   </td>
+   <td>
+s * 1000 + r * 500
+   </td>
+   <td>
+s * 1000
+   </td>
+   <td>
+s * 1000 + r * 1000
+   </td>
+   <td>
+s * 1000 - r * 1000
+   </td>
+  </tr>
+  <tr>
+   <td>
+0
+   </td>
+   <td>
+-500
+   </td>
+   <td>
+Only negative external 
+   </td>
+   <td>
+0
+   </td>
+   <td>
+r * 500
+   </td>
+   <td>
+0
+   </td>
+   <td>
+0
+   </td>
+   <td>
+0
+   </td>
+  </tr>
+  <tr>
+   <td>
+0
+   </td>
+   <td>
+500
+   </td>
+   <td>
+Only positive external
+   </td>
+   <td>
+0
+   </td>
+   <td>
+0
+   </td>
+   <td>
+0
+   </td>
+   <td>
+r * 500
+   </td>
+   <td>
+0
+   </td>
+  </tr>
+  <tr>
+   <td>
+0
+   </td>
+   <td>
+-500, 1000
+   </td>
+   <td>
+Negative and positive external
+   </td>
+   <td>
+0
+   </td>
+   <td>
+r * 500
+   </td>
+   <td>
+0
+   </td>
+   <td>
+r * 1000
+   </td>
+   <td>
+0
+   </td>
+  </tr>
+</table>
+
+
+
+
+#### Note
+* All 7 possible combination of position units that a given component can have are covered in the above table.
+    * Default: 0, External: n, p, np (3 cases)
+    * Default: 1, External:  0, n, p, np (4 cases)
+* Multiple negative external positions are the same as a single negative external position because we are summing up all the negative external positions in `DebtIssuanceModule#_getTotalIssuanceUnits`.
+* Multiple positive external positions are the same as a single positive external position because we are summing up all the positive external positions in `DebtIssuanceModule#_getTotalIssuanceUnits`.
+
+
+#### Note (for redeeming): 
+
+* SetToken's are burned at the very beginning of the DebtIssuanceModule#redeem() function, hence in columns 5,6,7,8 (count starting with 1) of the above table, setToken.totalSupply() is `s - r'`, where `r'` is the redeem quantity **without** fees.
+* After column 8, fees are minted to the fee receiver and setToken’s total supply becomes `s - r`.
+
+## Feasibility Analysis
+
+### Option 1:
+
+Instead of over-optimizing our checks to check for balances which effectively prevent both undercollaterlization and overcollaterlization. We can just have checks to prevent undercollaterlization. The change is contained within the smart contracts and doesn't bubble up the stack leading to bigger changes in our backend.
+
+The proposed checks for undercollateralization are performed right after the transfer happens, so that we revert immediately and save gas. When token is transferred in, we wish to check the right amount of equity/debt was passed in before we try to transfer it out to external protocols. When token is transferred out, we validate that we are not undercollateralized after the transfer.
+
+### Proposed checks during Issuance
+* In `_resolveEquityPositions` we can validate balances after the equity is transferred but before the external position hooks are called, because
+* Proposed check:
+
+```javascript
+// pseudocode
+uint256 defaultPositionUnit = setToken.getDefaultPositionRealUnit(component)
```

Comment:
I think we can optimize this by not doing as many state reads and external calls. We already know the total amount of equity that should be transferred because that's calculated in `_getTotalIssuanceUnits`. Combining with fetching the Default Position we should be able to back out what the External Equity is. One thing to consider is if there are any rounding errors we need to be aware of. I think what you have here works but it starts increasing costs quote a bit

Reply:
Yes, this was just a non-optimized pseudocode primarily added to show how the final check is derived and how each term of the final equations is calculated from fundamental premises.